text/0000-lazy-context-propagation.md

```diff
@@ -0,0 +1,106 @@
+- Start Date: 2019-06-19
+- RFC PR: (leave this empty)
+- React Issue: (leave this empty)
+
+# Summary
+
+This RFC describes a new Context Propagation implementation which has improved performance characteristics in certain cases by using the work React is already going to do before resorting to explicit context dependencies searches. This lazy implementation also has qualitative differences from the current eager approach which may allow new kinds of context update bailouts to be implemented
+
+It does not modify the public API of React in any way.
+
+# Motivation
+
+The eager context propagation in place today is executed per ContextProvider. When a ContextProvider is updating to a new value every constituent fiber has its context dependencies checked against the updating Context and work is scheduled if necessary
+
+This approach has two direct behaviors that can be optimized. First, for any given render, each context with a changed value will result in approximately the entire sub-tree for that context to be visited. In a simplified scenario, with ContextProviders at the top of the tree, if `N` contexts are updating then the fiber tree will be traversed during propagation exactly `N` times.

```

Comment:
elsewhere I refer to the visited sub-tree as a constituent sub-tree. the idea is that during current propagation you will not propagate through ContextProviders of the same type because deeper fibers are not affected by *this* instances value. So while propagation visits the entire sub-tree if there are no nested providers it does not do so when there are